### PR TITLE
feat(select):  add element customization

### PR DESCRIPTION
**Pull request by andipants12**
**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: packages/paste-core/components/select/stories/select.stories.tsx [hunk 1]
```diff
@@ -726,3 +728,89 @@ export const SelectOverflowLongValue = (): React.ReactNode => {
...
+        },
+        SELECT_ICON_WRAPPER: {
+          transform: 'rotate(90deg) translateX(-50%) translateY(-20%)',
+        },
+        CUSTOM_SELECT_ICON: {
+          color: 'colorTextInverseWeak',
+        },
+      }}
+    >
+      <Box maxWidth="size40" paddingX="space40" paddingY="space80">
```
Comment (andipants12) [inline packages/paste-core/components/select/stories/select.stories.tsx:779]: FYI I didn't create a component to share between the tests and stories since I did not want to create scope creep from modifying the unit tests.

File: .changeset/empty-spiders-rest.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+'@twilio-paste/input-box': patch
+'@twilio-paste/core': patch
+---
+
+[InputBox] Remove duplicative prefix in chevron icon wrapper element name, and update name too better reflect the component's purpose.
```
Comment (gloriliale) [inline .changeset/empty-spiders-rest.md:6]:
```suggestion
[InputBox] Remove duplicative prefix in chevron icon wrapper element name, and update name to better reflect the component's purpose.
```

File: packages/paste-core/components/input-box/src/InputChevronWrapper.tsx [hunk 3]
```diff
@@ -21,7 +21,7 @@ export const InputChevronWrapper = React.forwardRef<HTMLDivElement, InputChevron
       top="50%"
       transform="translateY(-50%)"
       zIndex="zIndex10"
-      element={`${element}_CHEVRON_WRAPPER`}
+      element={element}
```
Comment (gloriliale) [inline packages/paste-core/components/input-box/src/InputChevronWrapper.tsx:24]: This was a good catch! Especially when the term is already so long 😅

File: packages/paste-core/components/select/__tests__/option.test.tsx [hunk 4]
```diff
@@ -1,14 +1,19 @@
...
 import axe from '../../../../../.jest/axe-helper';
 import {Option} from '../src';
 import {createAttributeMap} from '../test-utils';
 
+expect.extend(matchers);
+
 interface MockOptionProps extends React.OptionHTMLAttributes<HTMLOptionElement> {
   suffix?: string;
   value?: string;
+  element?: string;
```
Comment (gloriliale) [inline packages/paste-core/components/select/__tests__/option.test.tsx:16]: Is it possible to use non-mocked props for this? or do we want them mocked?

File: packages/paste-core/components/select/__tests__/option.test.tsx [hunk 5]
```diff
@@ -74,4 +79,63 @@ describe('Form | Option', () => {
...
+            OPTION_GROUP: {
+              color: 'colorText',
+            },
+          }}
+        >
+          <MockOption suffix="custom-styles" />
+        </CustomizationProvider>
+      );
+
+      expect(screen.getByTestId('option-custom-styles')).toHaveStyleRule('color', 'rgb(3,11,93)');
```
Comment (gloriliale) [inline packages/paste-core/components/select/__tests__/option.test.tsx:117]: It might make sense to test for both of the stylings included in the custom theme, to make sure that both OPTION and OPTION_GROUP are behaving as expected. Thoughts?

Comment (gloriliale) [reply packages/paste-core/components/select/__tests__/option.test.tsx:117]: Oh I think I see that this is being tested in the next spec, got it!

File: packages/paste-core/components/select/src/types.ts [hunk 6]
```diff
@@ -1 +1,5 @@
+import type {BoxElementProps} from '@twilio-paste/box';
```
Comment (gloriliale) [inline packages/paste-core/components/select/src/types.ts:1]: this is cool, should this be our preferred way of adding Element into our component types?

**Review by gloriliale — APPROVED** [review]
This looks customizable to me! Great job catching the places where we could make improvements in the other packages. Is the new Element type in types one that we can now use for customization anywhere??

File: packages/paste-core/components/input-box/__tests__/input-chevron-wrapper.test.tsx [hunk 7]
```diff
@@ -8,13 +8,13 @@ expect.extend(matchers);
 
 describe('HTML attributes', () => {
   it('should set a element data attribute for InputChevronWrapper', () => {
-    render(<InputChevronWrapper element="INPUT">input-chevron</InputChevronWrapper>);
-    expect(screen.getByText('input-chevron').getAttribute('data-paste-element')).toEqual('INPUT_CHEVRON_WRAPPER');
+    render(<InputChevronWrapper element="INPUT_ICON_WRAPPER">input-chevron</InputChevronWrapper>);
+    expect(screen.getByText('input-chevron').getAttribute('data-paste-element')).toEqual('INPUT_ICON_WRAPPER');
```
Comment (SiTaggart) [inline packages/paste-core/components/input-box/__tests__/input-chevron-wrapper.test.tsx:12]: I would like to keep the element name as `x_CHEVRON_WRAPPER` in these type of inputs. Mainly because it's not just the input icon, it's specifically the input chevron, the select chevron, the combobox chevron, etc

File: packages/paste-core/components/select/__tests__/select.test.tsx [hunk 8]
```diff
@@ -171,13 +185,229 @@ describe('Form | Select', () => {
...
+      );
+    });
+  });
+
+  describe('Customization', () => {
+    const CustomizationWrapper: React.FC<{children: React.ReactNode}> = ({children}) => (
+      <CustomizationProvider
+        // @ts-expect-error global test variable
+        theme={TestTheme}
+        elements={{
```
Comment (nkrantz) [inline packages/paste-core/components/select/__tests__/select.test.tsx:229]: Ah now I see you're testing a lot of style rules down here in Select, nice!

**Review by nkrantz — APPROVED** [review]
Nice work Andi!

File: .changeset/five-lobsters-juggle.md [hunk 9]
```diff
@@ -0,0 +1,6 @@
+---
+'@twilio-paste/combobox': patch
+'@twilio-paste/core': patch
+---
+
+[Combobox]: Small fix for element naming pattern.
```
[the remark this answers is not in the document]
Comment (SiTaggart) [reply .changeset/five-lobsters-juggle.md:6]: It matters, the changelog this adds to is public record. It should read to inform the consumer of the library what happened

File: packages/paste-core/components/select/__tests__/option.test.tsx [hunk 10]
```diff
@@ -74,4 +79,63 @@ describe('Form | Option', () => {
...
+    });
+
+    it('should add custom styles to Option with a custom element data attribute', () => {
+      testRender(
+        <CustomizationProvider
+          // @ts-expect-error global test variable
+          theme={TestTheme}
+          elements={{
+            OPTION: {
+              color: 'colorTextLinkDarker',
```
[the remark this answers is not in the document]
Comment (SiTaggart) [reply packages/paste-core/components/select/__tests__/option.test.tsx:127]: Tricky one, but in these cases what we're testing is that the ability to override or add styles to the target element node. We're also testing the name too as that's a public API. So the element name can't change, and the functionality targeting that name can't change.

In the box and text utils where we do the style merging is where we do a little more specific testing of the style props. 

But even then we're reliant on styled system and the css method doing their part to merge the styles into emotion. Which also have their own tests.

So, the end of my long story is each layer of the stack has its own tests, and up here at the top of the stack we don't necessarily need to retest things that are happening down the stack. So testing one style property is perfectly fine and should provide the coverage we need.

File: .changeset/five-lobsters-juggle.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (andipants12) [reply .changeset/five-lobsters-juggle.md:6]: Yeah I have a todo list item to look up some examples of really good change log notes because my drafts definitely leave room for improvement. Will push an update to this.

File: packages/paste-core/components/input-box/__tests__/input-chevron-wrapper.test.tsx (hunk 7 above)
Comment (andipants12) [reply packages/paste-core/components/input-box/__tests__/input-chevron-wrapper.test.tsx:12]: Okey, that's fair. I will update 😸 I'm assuming the same for `Select` then too @SiTaggart?

File: packages/paste-core/components/select/__tests__/option.test.tsx (hunk 4 above)
Comment (andipants12) [reply packages/paste-core/components/select/__tests__/option.test.tsx:16]: I should probably rename this component function here, as it's not actually mocking anything, rather just creating a render of the component that we want to test configured for the test. 

That said, the interface for that component's props is called `MockOptionProps` because it's using a naming convention where the props interface is named by simply adding `Props` as a suffix to the component name.

Does that answer the question? Would renaming the props interface here make this more clear?
Would renaming the component `TestOption` and the props interface `TestOptionProps` make this more clear?
@gloriliale

File: packages/paste-core/components/select/__tests__/optiongroup.test.tsx [hunk 11]
```diff
@@ -1,12 +1,17 @@
...
+import {CustomizationProvider} from '@twilio-paste/customization';
 // @ts-ignore typescript doesn't like js imports
 import axe from '../../../../../.jest/axe-helper';
 import {OptionGroup, Option} from '../src';
 import {createAttributeMap} from '../test-utils';
 
+expect.extend(matchers);
+
 interface MockOptionGroupProps extends React.OptgroupHTMLAttributes<HTMLOptGroupElement> {
+  element?: string;
```
[the remark this answers is not in the document]
Comment (andipants12) [reply packages/paste-core/components/select/__tests__/optiongroup.test.tsx:14]: Not anything technically significant. I decided to simply declare the `element` type here since the rest of the "mock" test component interfaces are defined that way. Do you think this creates a level of confusion?

File: packages/paste-core/components/select/src/types.ts (hunk 6 above)
Comment (andipants12) [reply packages/paste-core/components/select/src/types.ts:1]: Yeah I've seen this done a few places in slightly different ways. Makes me wonder if it's worth creating a type in the box component package that is just `Element`

`export type Element = BoxElementProps['element'];`

Functionally it's the same as declaring this in each file, or skipping the declaration and just using the definition, but I think it might provide a layer of consistency and clarity. @SiTaggart @TheSisb Curious what you two think. 👀

File: packages/paste-core/components/input-box/__tests__/input-chevron-wrapper.test.tsx (hunk 7 above)
Comment (SiTaggart) [reply packages/paste-core/components/input-box/__tests__/input-chevron-wrapper.test.tsx:12]: Yup

File: .changeset/five-lobsters-juggle.md [hunk 12]
```diff
@@ -0,0 +1,6 @@
+---
+'@twilio-paste/combobox': patch
+'@twilio-paste/core': patch
+---
+
+[Combobox]: Update customization tests and stories to use more explicit a more explicit element name for the chevron wrapper. No consumer change required.
```
Comment (SiTaggart) [inline .changeset/five-lobsters-juggle.md:6]: Are this and input-box changelogs 100% accurate now?

**Review by SiTaggart — APPROVED** [review]
Outside of the question of changelogs, I think we're good

File: .changeset/five-lobsters-juggle.md [hunk 13]
```diff
@@ -0,0 +1,6 @@
+---
+'@twilio-paste/combobox': patch
+'@twilio-paste/core': patch
+---
+
+[Combobox]: Add suffix to Chevron Icon wrapper. This ensures that the element string value for the combobox's chevron wrapper will always include the suffix "_CHEVRON_WRAPPER", and use either the default element name, or the consumer provided custom element name. No consumer change required.
```
Comment (andipants12) [inline .changeset/five-lobsters-juggle.md:6]: @SiTaggart I just updated this one, because I think the previous change log wasn't super descriptive of the actual change.
